File: Solving/18877/18877.py
# Problem No.: 18877
# Solver:      Jinmin Goh
# Date:        20200709
# URL: https://www.acmicpc.net/problem/18877

import sys

def main():
    n, m = map(int, input().split())
    area = []
    grassLen = 0
    for _ in range(m):
        area.append(tuple(map(int, sys.stdin.readline().split())))
        grassLen += area[-1][1] - area[1][0] + 1
    area.sort()

    if grassLen == n:
        ans = 10 ** 18
        for i in range(m):
            if area[i][1] - area[i][0] > 0:
                ans = 0
                break
            elif i > 0:
                ans = min(ans, area[i][0] - area[i - 1][1])
        print(ans)
        return
    

    

    return

if __name__ == "__main__":
    main()

File: Solved/18877/18877.cpp
// Problem No.: 18877
// Solver:      Jinmin Goh
// Date:        20200820
// URL: https://www.acmicpc.net/problem/18877

#include <cstring>
#include <iostream>
#include <cstdlib>
#include <list>
#include <array>
#include <atomic>
#include <algorithm>
#include <deque>
#include <iterator>
#include <map>
#include <numeric>
#include <queue>
#include <set>
#include <stack>
#include <string>
#include <valarray>
#include <vector>
#include <tuple>
#include <unordered_map>
#include <unordered_set>
using namespace std;

vector<pair<long long, long long>> interval;

int main() {
    long long n, m;
    scanf("%lld %lld", &n, &m);
    for(int i = 0; i < m; i++) {
        long long a, b;
        scanf("%lld %lld", &a, &b);
        interval.push_back(make_pair(a, b));
    }

    sort(interval.begin(), interval.end());
    long long lo = 0, hi = 1e18 + 1;

    while(lo < hi) {
        long long mid = (lo + hi + 1) / 2;
        long long prev = -1e18 - 1;
        int cnt = 0;
        for (auto& i : interval) {
            while (max(prev + mid, i.first) <= i.second) {
                prev = max(prev + mid, i.first);
                cnt++;
                if (cnt >= n) break;
            }
            if (cnt >= n) break;
        }

        if(cnt >= n) {
            lo = mid;
        }
        else {
            hi = mid - 1;
        }
    }
    printf("%lld", lo);
    return 0;
}